internal/dot/graph.go

```diff
@@ -252,6 +318,87 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) PruneSuccess() {
+	dg.pruneCtors()
+	dg.pruneGroups()
+}
+
+// Remove constructors from the graph that do not have failing Results.
+func (dg *Graph) pruneCtors() {
+	var pruned []*Ctor
+	for _, c := range dg.Ctors {
+		if _, ok := dg.Failed.ctors[c.ID]; ok {
+			pruned = append(pruned, c)
+		} else {
+			dg.pruneCtorParams(c)
+			dg.pruneGroupResults(c)
+			delete(dg.ctorMap, c.ID)
+		}
+	}
+
+	dg.Ctors = pruned
+}
+
+// pruneGroups removes groups from the graph that do not have failing results.
+func (dg *Graph) pruneGroups() {
+	var pruned []*Group
+	for _, g := range dg.Groups {
+		k := g.groupKey()
+		if _, ok := dg.Failed.groups[k]; ok {
+			pruned = append(pruned, g)
+		} else {
+			delete(dg.groupMap, k)
+		}
+	}
+	dg.Groups = pruned
+
+	dg.pruneCtorGroupParams()
+}
+
+// pruneCtorParams removes constructor results that are still referenced in the Params of constructors that consume those results.
+// If the results in the constructor are found in the params of a consuming constructor that result should be removed.
+func (dg *Graph) pruneCtorParams(c *Ctor) {
+	for _, r := range c.Results {
+		if ctors, ok := dg.paramCtorMap[r.nodeKey()]; ok {
+			for _, ctor := range ctors {
+				ctor.removeParams(r)
+			}
+		}
+	}
+}
+
+// pruneCtorGroupParams removes constructor results that are still referenced in the GroupParams of constructors that consume those results.
+func (dg *Graph) pruneCtorGroupParams() {
+	for _, c := range dg.Ctors {
+		var pruned []*Group
+		for _, gp := range c.GroupParams {
+			k := gp.groupKey()
+			if _, ok := dg.groupMap[k]; ok {
+				pruned = append(pruned, gp)
+			}
+		}
+		c.GroupParams = pruned
+	}
+}
+
+// pruneGroupResults removes constructor results that are still referenced in the Group object that contains that result.
+func (dg *Graph) pruneGroupResults(c *Ctor) {
```

Comment:
we should probably document which constructor `c` is here and in the other function.

Reply:
I added documentation in the comments re: what the c argument is. Is there a better pattern for parameter documentation? I tried to look online, but there didn't seem to be an idiomatic way.